File: frontend/src/services/graph.ts
/**
 * AI4EDU 知识图谱 API
 */
import api from './api'

// ============ 类型定义 ============

export interface SquareStat {
  id: string
  name: string
  icon: string
  color: string
  node_count: number
  completeness: number
  misconception_count: number
}

export interface KnowledgeNode {
  id: string
  name: string
  subject?: string
  description?: string
  cognitive_level?: Record<string, number>
  has_misconception?: boolean
  misconceptions?: Misconception[]
  [key: string]: unknown
}

export interface GraphLink {
  source: string
  target: string
  type: string
  label: string
  strength?: number
  strength_label?: '强' | '中' | '弱'
  is_cross?: boolean
}

export interface NeighborData {
  nodes: KnowledgeNode[]
  links: GraphLink[]
}

export interface CognitiveGoal {
  dimensions: string[]
  dimension_keys?: string[]
  values: number[]
  subject_avg?: number[] | null
  node_name: string
}

// ========== Misconception 类型 ==========

export interface Misconception {
  mc_id: string
  misconception: string
  correction: string
  topic: string
  keywords: string[]
  source: 'teacher' | 'ai' | 'system'
  annotated_by: number | null
  annotated_at: string
  confidence: number
}

export interface MisconceptionInput {
  misconception: string
  correction: string
  topic: string
  keywords: string[]
}

export interface MisconceptionSuggestion {
  misconception: string
  correction: string
  topic: string
  keywords: string[]
  subject: string
  confidence: number
}

export interface AutoSuggestResponse {
  suggestions: MisconceptionSuggestion[]
  total: number
}

// ========== 跨学科图谱类型 ==========

export interface CrossSubjectNode {
  id: string
  name: string
  subject: string
  description?: string | null
  has_misconception: boolean
  degree: number
}

export interface CrossSubjectLink {
  source: string
  target: string
  type: string
  label: string
  is_cross: boolean
  strength: number
  strength_label: '强' | '中' | '弱'
}

export interface CrossSubjectStats {
  total_nodes: number
  total_links: number
  cross_links: number
  avg_strength: number
  subject_distribution: Record<string, number>
  top_cross_pairs?: Array<Record<string, unknown>>
}

export interface CrossSubjectResponse {
  nodes: CrossSubjectNode[]
  links: CrossSubjectLink[]
  stats: CrossSubjectStats
}

// ========== 任务类型 ==========

export interface NodeTask {
  task_id: string
  name: string
  type: 'learning' | 'review' | 'diagnosis'
  status: 'pending' | 'in_progress' | 'completed'
  due_date: string | null
  source: 'course' | 'diagnosis' | 'teacher'
  description?: string
  assigned_by?: number
}

export type GraphChangeType =
  | 'overview'
  | 'cognitive'
  | 'relationship'
  | 'relationship_delete'
  | 'recommendation'
  | 'recommendation_delete'
  | 'resource_link'
  | 'resource_update'
  | 'resource_unlink'

export interface GraphChangeRequest {
  id: string
  node_id: string
  change_type: GraphChangeType
  payload: Record<string, unknown>
  status: 'pending' | 'approved' | 'rejected'
  submitted_by: number
  reviewer_id?: number
  review_comment?: string
  created_at: string
}

export interface GraphChangeResult {
  status: 'approved' | 'pending'
  applied?: unknown
  id?: string
}

// ============ API 方法 ============

export const graphApi = {
  /** 获取图谱广场统计 */
  async getSquareStats(): Promise<SquareStat[]> {
    const response = await api.get('/graphs/square')
    return response.data
  },

  /** 获取节点详情 */
  async getNodeDetail(nodeId: string): Promise<KnowledgeNode> {
    const response = await api.get(`/graphs/nodes/${nodeId}`)
    return response.data
  },

  /** 获取邻居节点 */
  async getNeighbors(nodeId: string, depth = 1, limit = 50): Promise<NeighborData> {
    const response = await api.get(`/graphs/nodes/${nodeId}/neighbors`, {
      params: { depth, limit },
    })
    return response.data
  },

  /** 获取节点关联资源 */
  async getNodeResources(nodeId: string): Promise<unknown[]> {
    const response = await api.get(`/graphs/nodes/${nodeId}/resources`)
    return response.data
  },

  /** 获取推荐节点 */
  async getRecommendations(nodeId: string, limit = 10): Promise<KnowledgeNode[]> {
    const response = await api.get(`/graphs/nodes/${nodeId}/recommendations`, {
      params: { limit },
    })
    return response.data
  },

  /** 获取认知目标 */
  async getCognitiveGoals(nodeId: string, includeAvg = false): Promise<CognitiveGoal> {
    const response = await api.get(`/graphs/nodes/${nodeId}/cognitive`, {
      params: { include_avg: includeAvg },
    })
    return response.data
  },

  /** 创建知识节点 */
  async createNode(data: { id: string; name: string; subject: string; description?: string }): Promise<KnowledgeNode> {
    const response = await api.post('/graphs/nodes', null, { params: data })
    return response.data
  },

  /** 更新知识节点 */
  async updateNode(nodeId: string, data: Record<string, unknown>): Promise<KnowledgeNode> {
    const response = await api.put(`/graphs/nodes/${nodeId}`, null, { params: data })
    return response.data
  },

  /** 创建节点关系 */
  async createRelationship(fromId: string, toId: string, relType = 'RELATED', label?: string): Promise<unknown> {
    const response = await api.post(`/graphs/nodes/${fromId}/link/${toId}`, null, {
      params: { rel_type: relType, label },
    })
    return response.data
  },

  /** 删除节点关系 */
  async deleteRelationship(fromId: string, toId: string, relType = 'RELATED'): Promise<void> {
    await api.delete(`/graphs/nodes/${fromId}/link/${toId}`, {
      params: { rel_type: relType },
    })
  },

  /** 搜索知识节点 */
  async searchNodes(q: string, subject?: string, limit = 20): Promise<KnowledgeNode[]> {
    const response = await api.get('/graphs/search', {
      params: { q, subject, limit },
    })
    return response.data
  },

  // ========== Misconception API ==========

  /** 获取节点误解标注列表 */
  async getMisconceptions(nodeId: string): Promise<Misconception[]> {
    const response = await api.get(`/graphs/nodes/${nodeId}/misconceptions`)
    return response.data
  },

  /** 教师添加误解标注 */
  async addMisconception(nodeId: string, data: MisconceptionInput): Promise<Misconception> {
    const response = await api.post(`/graphs/nodes/${nodeId}/misconceptions`, data)
    return response.data
  },

  /** 教师编辑误解标注 */
  async updateMisconception(nodeId: string, mcId: string, data: Partial<MisconceptionInput>): Promise<Misconception> {
    const response = await api.put(`/graphs/nodes/${nodeId}/misconceptions/${mcId}`, data)
    return response.data
  },

  /** 教师删除误解标注 */
  async deleteMisconception(nodeId: string, mcId: string): Promise<void> {
    await api.delete(`/graphs/nodes/${nodeId}/misconceptions/${mcId}`)
  },

  /** AI辅助标注建议 */
  async suggestMisconceptions(nodeId: string): Promise<AutoSuggestResponse> {
    const response = await api.post(`/graphs/nodes/${nodeId}/misconceptions/auto-suggest`)
    return response.data
  },

  // ========== 任务 API ==========

  /** 获取节点关联任务 */
  async getNodeTasks(nodeId: string): Promise<NodeTask[]> {
    const response = await api.get(`/graphs/nodes/${nodeId}/tasks`)
    return response.data
  },

  /** Submit a graph edit. Teacher/admin edits apply immediately; student edits await review. */
  async submitChange(
    nodeId: string,
    changeType: GraphChangeType,
    payload: Record<string, unknown>,
  ): Promise<GraphChangeResult> {
    const response = await api.post(`/graphs/nodes/${nodeId}/changes`, {
      change_type: changeType,
      payload,
    })
    return response.data
  },

  async getChangeRequests(
    nodeId?: string,
    status: 'pending' | 'approved' | 'rejected' = 'pending',
  ): Promise<GraphChangeRequest[]> {
    const response = await api.get('/graphs/change-requests', {
      params: { node_id: nodeId, status },
    })
    return response.data
  },

  async reviewChange(requestId: string, approved: boolean, comment?: string): Promise<void> {
    await api.post(`/graphs/change-requests/${requestId}/review`, { approved, comment })
  },

  async createNodeTask(
    nodeId: string,
    data: { name: string; description?: string; type: NodeTask['type']; due_date?: string },
  ): Promise<NodeTask> {
    const response = await api.post(`/graphs/nodes/${nodeId}/tasks`, data)
    return response.data
  },

  // ========== 跨学科图谱 API ==========

  /** 获取跨学科关联图谱 */
  async getCrossSubjectGraph(subjects: string[], minStrength = 0.0, maxNodes = 100): Promise<CrossSubjectResponse> {
    const response = await api.get('/graphs/cross-subject', {
      params: {
        subjects: subjects.join(','),
        min_strength: minStrength,
        max_nodes: maxNodes,
      },
    })
    return response.data
  },
}

File: backend/app/schemas/graph.py
"""
AI4Edu 知识图谱模块 Pydantic Schema 集中定义
包含 Misconception、CrossSubject、Task、CognitiveGoal 等所有图谱相关 Schema
"""
from datetime import datetime
from typing import Any, Dict, List, Optional, Literal

from pydantic import BaseModel, Field


# ========== Misconception ==========

class MisconceptionItem(BaseModel):
    """误解标注项（读取/响应）"""

    mc_id: str
    misconception: str
    correction: str
    topic: str
    keywords: List[str] = Field(default_factory=list)
    source: Literal["teacher", "ai", "system"]
    annotated_by: Optional[int] = None
    annotated_at: datetime
    confidence: float = Field(default=1.0, ge=0.0, le=1.0)


class MisconceptionCreate(BaseModel):
    """创建误解标注（请求体）"""

    misconception: str = Field(..., min_length=1, max_length=500)
    correction: str = Field(..., min_length=1, max_length=2000)
    topic: str = Field(..., min_length=1, max_length=200)
    keywords: List[str] = Field(default_factory=list, max_length=10)


class MisconceptionUpdate(BaseModel):
    """更新误解标注（请求体，所有字段可选）"""

    misconception: Optional[str] = Field(None, max_length=500)
    correction: Optional[str] = Field(None, max_length=2000)
    topic: Optional[str] = Field(None, max_length=200)
    keywords: Optional[List[str]] = None


class MisconceptionSuggestion(BaseModel):
    """AI 辅助标注建议"""

    misconception: str
    correction: str
    topic: str
    keywords: List[str]
    subject: str
    confidence: float = Field(ge=0.0, le=1.0)


class AutoSuggestResponse(BaseModel):
    """AI 辅助标注建议响应"""

    suggestions: List[MisconceptionSuggestion]
    total: int


# ========== 广场统计（扩展） ==========

class SquareStatItem(BaseModel):
    """学科广场统计项"""

    id: str
    name: str
    icon: str
    color: str
    node_count: int
    completeness: float
    misconception_count: int = 0


# ========== 跨学科图谱 ==========

class CrossSubjectNode(BaseModel):
    """跨学科图谱节点"""

    id: str
    name: str
    subject: str
    description: Optional[str] = None
    has_misconception: bool = False
    degree: int = 0


class CrossSubjectLink(BaseModel):
    """跨学科图谱关系"""

    source: str
    target: str
    type: str
    label: str
    is_cross: bool
    strength: float = Field(ge=0.0, le=1.0)
    strength_label: Literal["强", "中", "弱"]


class CrossSubjectStats(BaseModel):
    """跨学科图谱统计"""

    total_nodes: int
    total_links: int
    cross_links: int
    avg_strength: float
    subject_distribution: Dict[str, int]
    top_cross_pairs: List[Dict[str, Any]] = Field(default_factory=list)


class CrossSubjectResponse(BaseModel):
    """跨学科图谱响应"""

    nodes: List[CrossSubjectNode]
    links: List[CrossSubjectLink]
    stats: CrossSubjectStats


# ========== 任务 ==========

class NodeTask(BaseModel):
    """节点关联任务"""

    task_id: str
    name: str
    type: Literal["learning", "review", "diagnosis"]
    status: Literal["pending", "in_progress", "completed"]
    due_date: Optional[datetime] = None
    source: Literal["course", "diagnosis"]


# ========== 认知目标（扩展） ==========

class CognitiveGoalResponse(BaseModel):
    """认知目标雷达图数据"""

    dimensions: List[str]
    dimension_keys: List[str]
    values: List[float]
    subject_avg: Optional[List[float]] = None
    node_name: str


# ========== Collaborative graph editing ==========

class GraphChangeCreate(BaseModel):
    """A graph change that is applied directly or sent for teacher review."""

    change_type: Literal[
        "overview",
        "cognitive",
        "relationship",
        "relationship_delete",
        "recommendation",
        "recommendation_delete",
        "resource_link",
        "resource_update",
        "resource_unlink",
    ]
    payload: Dict[str, Any] = Field(default_factory=dict)


class GraphReviewAction(BaseModel):
    approved: bool
    comment: Optional[str] = Field(None, max_length=500)


class GraphTaskCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    description: str = Field(default="", max_length=2000)
    type: Literal["learning", "review", "diagnosis"] = "learning"
    due_date: Optional[datetime] = None
